Core/src/org/sleuthkit/autopsy/guiutils/WrapLayout.java

```diff
@@ -0,0 +1,607 @@
+     *                      baseline.
+     * @param rowHeight     The height of the row.
+     * @param itemHeight    The height of the item.
+     *
+     * @return The top y position of the component.
+     */
+    private int getComponentY(int rowY, boolean alignBaseline, int rowHeight, int itemHeight) {
+        return alignBaseline
+                ? rowY + rowHeight - itemHeight
+                : rowY;
+    }
+
+    /**
+     * * Determines the subcomponent's x position.
+     *
+     * @param leftX          The leftmost position a component can be placed.
+     * @param rightX         The rightmost position a component can be placed.
+     * @param ltr            If the components should be laid out left to right.
+     * @param xPos           The x position of the component (if left to right,
+     *                       how far from leftX; otherwise how far from rightX).
+     * @param componentWidth The component's width.
+     *
+     * @return The component's left x position.
+     */
+    private int getComponentX(int leftX, int rightX, boolean ltr, int xPos, int componentWidth) {
+        return ltr ? leftX + xPos : rightX - xPos - componentWidth;
+    }
+
+    /**
+     * Sets a subcomponent's size to preferred size and sets the (x,y) position
+     * for the component.
+     *
+     * @param comp          The component.
+     * @param alignBaseline Whether this component should be aligned on the
+     *                      baseline.
+     * @param ltr           If the components should be laid out left to right.
+     * @param rowY          The top y position of the row.
+     * @param rowHeight     The height of the row.
+     * @param leftX         The leftmost position a component can be placed.
+     * @param rightX        The rightmost position a component can be placed.
+     * @param xPos          The x position of the component (if left to right,
+     *                      how far from leftX; otherwise how far from rightX).
+     *
+     * @return The width of the component.
+     */
+    private int setComponentDims(Component comp, boolean alignBaseline, boolean ltr, int rowY, int rowHeight, int leftX, int rightX, int xPos) {
+        Dimension d = comp.getPreferredSize();
+        comp.setSize(d);
+
+        int x = getComponentX(leftX, rightX, ltr, xPos, d.width);
+        int y = getComponentY(rowY, alignBaseline, rowHeight, d.height);
+        comp.setLocation(x, y);
+
+        return d.width;
+    }
+
+    /**
+     * Lays out the container. This method lets each
+     * <i>visible</i> component take its preferred size by reshaping the
+     * components in the target container and creating new rows.
+     *
+     * @param target The specified component being laid out.
+     */
+    @Override
+    public void layoutContainer(Container target) {
+        synchronized (target.getTreeLock()) {
+            synchronized (this.oppositeAlignedItems) {
```

Comment:
getting locks inside locks seems like a recipe for problems to me but I will let @rcordovano  weigh in on this one.

Reply:
My aim was to synchronize on both items similar to https://stackoverflow.com/a/4604038 to get access to the tree lock to avoid problems with altering objects in a container during layout: https://docs.oracle.com/javase/7/docs/api/java/awt/Component.html#getTreeLock() as well as prevent access to the collection created for objects that will be oppositely aligned during layout operations.

Reply:
@gdicristofaro and @wschaeferB, nested locking is really risky and the consequence of a mistake, now or in maintenance, is a deadlock. Given that this code is UI code that will be thread-confined in the EDT, there should be no need for any synchronization at all in this class. Swing components get their thread safety by thread-confinement by design. Please remove all synchronization.